crates/orderbook/src/cow_subsidy.rs

```diff
@@ -0,0 +1,85 @@
+use std::{sync::Mutex, time::Duration};
+
+use anyhow::Result;
+use cached::{Cached, TimedSizedCache};
+use contracts::ERC20;
+use primitive_types::{H160, U256};
+
+const CACHE_SIZE: usize = 10000;
+const CACHE_LIFESPAN: Duration = Duration::from_secs(60 * 60);
+
+/// Check whether a user qualifies for an extra fee subsidy because they own enough cow token.
+#[async_trait::async_trait]
+pub trait CowSubsidy: Send + Sync + 'static {
+    async fn qualifies_for_subsidy(&self, user: H160) -> Result<bool>;
```

Comment:
Does it make sense for this component to be a simple `CowBalance` component to keep all subsidy logic in `fee.rs`?

The main reason I was suggesting this is that I'm 90% sure that we will want this to be tiered so we will need the balance amount and not a binary "has balance or not". Additionally, it feels a bit weird to me that the subsidy threshold is configured here and the subsidy amount is configured by the `MinFeeCalculating` implementation (i.e. the code could be a bit more colocated).

Reply:
Alternatively we can store the threshold and factor both in this component and have `fn cow_subsidy(&self, user: H160) -> Result<f64>`. One difference is that having all the cow subsidy related code is in one place then so that the balance fetching is not separate from the fee calculation. It could be good that this keeps the fee component simpler. This would be reflected in the cache too as at the moment the cache can store `bool` but in your example it would store `U256`.
I'm not sure what is better.

Reply:
Hmm, good point. I guess it does make sense to keep the fee component simpler. Then I would suggest `fn cow_subsidy(&self, user: H160) -> Result<f64>` so that COW related subsidy code is in one place.